### PR TITLE
feat(languages): sql

### PR DESCRIPTION
**Pull request by matoous**
WIP, needs some fixing

Screen Shot 2022-01-08 at 19 44 57
.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -198,3 +198,6 @@
 	path = helix-syntax/languages/tree-sitter-make
 	url = https://github.com/alemuller/tree-sitter-make
 	shallow = true
+[submodule "helix-syntax/languages/tree-sitter-sql"]
+	path = helix-syntax/languages/tree-sitter-sql
+	url = git@github.com:m-novikov/tree-sitter-sql.git
```
Comment (pickfire) [inline .gitmodules:203]:
```suggestion
	url = git@github.com:m-novikov/tree-sitter-sql.git
	shallow = true
	
```

File: languages.toml [hunk 2]
```diff
@@ -545,3 +545,13 @@ scope = "source.regex"
...
+
+[[language]]
+name = "sql"
+scope = "source.sql"
+roots = []
+file-types = ["sql"]
+injection-regex = "sql"
+comment-token = "--"
+indent = { tab-width = 2, unit = " " }
+
```
Comment (pickfire) [inline languages.toml:557]:
```suggestion
```

File: languages.toml [hunk 3]
```diff
@@ -545,3 +545,13 @@ scope = "source.regex"
...
 roots = []
+
+[[language]]
+name = "sql"
+scope = "source.sql"
+roots = []
+file-types = ["sql"]
+injection-regex = "sql"
+comment-token = "--"
+indent = { tab-width = 2, unit = " " }
```
Comment (pickfire) [inline languages.toml:556]: Hmm, SQL 2 spaces? I never used SQL in two spaces, I thought it's rare.

File: runtime/queries/sql/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  "LOCAL"
+  "SESSION"
+  "TABLE"
+  "TO"
+  "UNIQUE"
+  "UPDATE"
+  "USAGE"
+  "VALUES"
+  "WHERE"
+] @keyword
```
Comment (pickfire) [inline runtime/queries/sql/highlights.scm:101]: Would it be better to separate `CREATE`, `UPDATE`, `DELETE` as `keyword.operator`? Also, I don't see `DELETE`.

Comment (matoous) [reply runtime/queries/sql/highlights.scm:101]: I am not sure about having them as `keyword.operator`. Looking at other implementations (Intellihj, VScode) it doesn't seem that anyone is doing this. On the other hand, some implementations have for example different highlighting for logic operators (`OR`, `AND`).

**Comment by matoous** [comment]
Hmm, there also seems to be https://github.com/DerekStride/tree-sitter-sql that has more recent changes, same amount of stars as the repo used currently in this PR (https://github.com/m-novikov/tree-sitter-sql).

File: runtime/queries/sql/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+(function_call
+    function: (identifier) @function)
+
+(identifier) @variable
+
+[
+  (NULL)
+  (TRUE)
+  (FALSE)
```
Comment (archseer) [inline runtime/queries/sql/highlights.scm:13]: probably want to split these out as boolean

**Comment by jaredramirez** [comment]
I've set up sql support in a fork and started with https://github.com/m-novikov/tree-sitter-sql. After using it for a little while, I switched to https://github.com/DerekStride/tree-sitter-sql. IMO it's a nice grammar and you can write nicer queries/highlights.

**Comment by the-mikedavis** [comment]
I did some looking around and I don't see any SQL grammars that look very close to complete[^1]

I think the highlighting will be fairly buggy-looking for even DerekStride/tree-sitter-sql, but I also think that grammar is the pick of the litter with what's available currently. From what I understand, SQL is quite a large and nuanced language, so it's hard to pull off a grammar that handles everything that's possible with it.

The good news is injections should be easier than I was thinking with the rust-regex-style injection. Based on [this discussion](https://github.com/tree-sitter/tree-sitter/discussions/1577), you could write an injections.scm query like so:

```scm
((string) @injection.content
  (#match? @injection.content "^\s*(SELECT|CREATE|..)") ; .. etc
  (#set injection.language "sql"))
```

and that should be flexible enough to inject SQL into string literals where that's common (like go and rust afaik)

[^1]: I'm particularly fond of [this](https://github.com/tjdevries/tree-sitter-sql#status) grammar's bluntly honest "status" section :smile:.

**Comment by archseer** [comment]
> 1. I'm particularly fond of [this](https://github.com/tjdevries/tree-sitter-sql#status) grammar's bluntly honest "status" section smile. [](#user-content-fnref-1-4190c97b0c967e5813fc34019fb19321)

Looks like I wasn't the first to think of that 😄

**Comment by jaredramirez** [comment]
@matoous what's the status on this? I'm pretty excited for this to land as it's one of the last languages not already supported I used regularly and I'd love to switch to Helix as my daily editor. Is there anything I can do to help?

**Comment by matoous** [comment]
@jaredramirez Sorry, got a lot of work lately so I didn't have time to take a look yet. Currently, `highlight.scm` needs to be written given that there's no predefined one in https://github.com/DerekStride/tree-sitter-sql. You can either pick up where I left or I will take a look hopefully next week.

**Comment by jaredramirez** [comment]
Sorry for the delay in responding! Tested with Helix and it's all good. Thanks for checking!

**Comment by the-mikedavis** [comment]
superseded by #2097

**Pull request closed without merging**